### ISSUE TITLE
gfxrecon-replay should emit an identifiable command or block after restoring state

### ISSUE THREAD

**Issue by lunarpapillo** [LunarG/gfxreconstruct#726]
Right now, if `gfxrecon-replay` is used on a trimmed capture, and the output is captured again, information is lost: the replay operation will create a semi-random number of additional frames, which will then become part of the new capture, offsetting all following frames by a difficult-to-determine amount.

If `gfxrecon-replay` emitted some sort identifiable `NOP` command or block (with no impact on rendering, but that would be stored in a capture file if the capture facility were active), that information is preserved.

Then when replaying such a re-captured trimmed capture, parsing the special command or block could reset the frame count, effectively preventing recaptured state blocks from inadvertently drifting the frame count, and keeping the user from having to be responsible for detecting and compensating for such frame drift.

**Comment by panos-lunarg** [comment]
From what I can see the additional frames are generated when replaying the trimmed capture.
Don't fully understand the concept of the mechanism but it seems that the additional frames are generated in order to prepare the state of the swapchain images before starting the actual replay.

I **think** we can avoid capturing those extra frames by calling directly the driver's `vkQueuePresent`, bypassing gfxreconstruct (and the loader). Would that do the trick or do we really need to store that information in the new capture (as "NOP"s for example)?

**Comment by lunarpapillo** [comment]
> I **think** we can avoid capturing those extra frames by calling directly the driver's `vkQueuePresent`, bypassing gfxreconstruct (and the loader). Would that do the trick or do we really need to store that information in the new capture (as "NOP"s/special commands for example)?

Interesting... I could see that working.  But I'd imagine it might not be trivial to get to the driver directly, or even to figure out which driver (and device, if the driver supports more than one on the system) to use...

**Comment by panos-lunarg** [comment]
After experimenting a bit more with the concept I realized that bypassing GFXR layer/loader is not possible (or at least I couldn't find a way to make it work)

Instead I tried exposing two new functions from the recording layer. Those two functions are discoverable through `vkGetDeviceProcAddr` when the layer is enabled. When replaying, `gfxrecon-replay` will look for these two new functions. If the layer is enabled then they will be detected, otherwise `vkGetDeviceProcAddr` will return null pointers.

These two functions essentially tell the `CaptureManager` class to stop and resume recording respectively. So when  `gfxrecon-replay` is replaying a trimmed capture, the function that is responsible for setting up the swapchain images, and as a result offset the total frame count, will stop recording, do the setup and then resume recording.

I tried out the patch with vkCube. From what I have seen, capturing a replay of a trimmed capture, offsets the total frames by 6. The patch seems to handle that ok although it changes the (until now) "expected" output when taking screenshots. More specifically taking screenshots when replaying the capture of a trimmed replay:
- Without the patch, the first 6 screenshots are completely black (which are the 6 additional frames which the whole replay is offseted by).
- With the patch the black screenshots are gone. The first screenshot is the same as screenshot number 7 taken without the patch.

So the patch is going to introduce some breakage in the CI tests.

Created experimental PR #743 . Let's see if it has the desired behavior and then we can decide whether we agree with changes and formalize it (it will require some changes in the Vulkan generation scripts which I haven't figured out yet), or try a different approach.

**Comment by lunarpapillo** [comment]
> From what I have seen, capturing a replay of a trimmed capture, offsets the total frames by 6. The patch seems to handle that ok although it changes the (until now) "expected" output when taking screenshots.

That's nothing short of awesome!!! :tada:

> So the patch might introduce some breakage in the CI tests.

I actually suspect it won't.  The CI `trim-single` and `trim-range` exercises are both written to expect a variable number of unintended frames; I think they'll work fine with 0 unintended frames (though the test could could be simplified and made more precise as a second effort).

Currently recapture exercises are disabled for trimmed traces, because that exercise is _not_ state-frame-aware.  But because the exercises aren't being executed on trimmed traces, they won't fail.  And (in theory at least) we could re-enable them and expect them to pass now...

I'll schedule some tests.

**Comment by bradgrantham-lunarg** [comment]
PR #701 solved the problem of the initial empty images, so I'm going to lower the priority of this issue to P2.

However, there's still the issue of lost state when recapturing (replaying with capture).  I think what needs to happen is that metacommands from the initial capture file should be somehow preserved (through a command-line option?) in recapture into the resulting capture file.

**Comment by lunarpapillo** [comment]
> However, there's still the issue of lost state when recapturing (replaying with capture).

I thought that replaying with capture set state using normal Vulkan function calls, and that capturing them would capture the function calls used to set the state just as it would any other Vulkan calls...?   I may be confused, or may not understand exactly what the virtual swapchain did or did not do.

**Comment by bradgrantham-lunarg** [comment]
> I thought that replaying with capture set state using normal Vulkan function calls, and that capturing them would capture the function calls used to set the state just as it would any other Vulkan calls...?

Yeah, it will capture all the *Vulkan* calls, but we don't pass through and store any of the MetaCommand blocks...  I was just thinking that a capture of a replay of a trimmed capture doesn't preserve the fact that the original capture file was trimmed because we don't pass through the begin-state and end-state markers.  The new capture won't report it was trimmed in `gfxrecon-info`.  Maybe that's okay for now, thus the P2.
